synapse/lib/cell.py

```diff
@@ -4053,3 +4069,258 @@ async def _runReloadFunc(self, name, func):
+        '''
+        user = await self.auth.reqUser(useriden)
+        iden, token, shadow = await s_passwd.generateApiKey()
+        now = s_common.now()
+        kdef = {
+            'iden': iden,
+            'name': name,
+            'user': user.iden,
+            'created': now,
+            'updated': now,
+            'shadow': shadow,
+        }
+
+        if duration:
+            if duration < 1:
+                raise s_exc.BadArg(mesg='Duration must be equal or greater than 1', name='duration')
+            kdef['expires'] = now + duration
+
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:add', kdef)
+
+        logger.info(f'Created HTTP API Key {iden} for {user.name}, {name=}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return token, kdef
+
+    @s_nexus.Pusher.onPush('user:apikey:add')
+    async def _genUserApiKey(self, kdef):
+        iden = s_common.uhex(kdef.get('iden'))
+        user = s_common.uhex(kdef.get('user'))
+        self.slab.put(iden, user, db=self.apikeydb)
+        lkey = user + b'apikey' + iden
+        self.slab.put(lkey, s_msgpack.en(kdef), db=self.usermetadb)
+
+    async def getUserApiKey(self, iden):
+        '''
+        Get a user API key via iden.
+
+        Notes:
+            This contains the raw value. Callers are responsible for removing the ``shadow`` key.
+
+        Args:
+            iden (str): The key iden.
+
+        Returns:
+            dict: The key dictionary; or none.
+        '''
+        lkey = s_common.uhex(iden)
+        user = self.slab.get(lkey, db=self.apikeydb)
+        if user is None:
+            return None
+        buf = self.slab.get(user + b'apikey' + lkey, db=self.usermetadb)
+        if buf is None:  # pragma: no cover
+            logger.warning(f'Missing API key {iden} from user metadata for {s_common.ehex(user)}')
+            return None
+        kdef = s_msgpack.un(buf)  # This includes the shadow key
+        return kdef
+
+    async def listUserApiKeys(self, useriden):
+        '''
+        Get all the API keys for a user.
+
+        Args:
+            useriden (str): The user iden.
+
+        Returns:
+            list: A list of kdef values.
+        '''
+        user = await self.auth.reqUser(useriden)
+        vals = []
+        prefix = s_common.uhex(user.iden) + b'apikey'
+        for lkey, valu in self.slab.scanByPref(prefix, db=self.usermetadb):
+            kdef = s_msgpack.un(valu)
+            kdef.pop('shadow')
+            vals.append(kdef)
+            await asyncio.sleep(0)
+        return vals
+
+    async def getApiKeys(self):
+        '''
+        Get all API keys in the cell.
+
+        Yields:
+            tuple: Useriden, kdef values
```

Comment:
```suggestion
            tuple: Useriden, kdef values.
```

Reply:
This needs to be rewritten since we no longer include the useriden since that is in kdef.